File: implementations/ts/s57_e2e.test.ts
import test from 'node:test';
import assert from 'node:assert/strict';

import { H57Length } from './h57.js';
import { ID57Length } from './id57.js';
import { S57 } from './s57.js';

const enc = new TextEncoder();
const dec = new TextDecoder();

test('S57 e2e secure composition flow', () => {
  const s57 = new S57({
    server_secret_key: enc.encode('S57_SERVER_SECRET_KEY_MUST_BE_LONG_1234567890'),
    environment_salt: enc.encode('staging-v1'),
    key_id: 12
  });

  const payload = enc.encode('e2e-s57-payload');

  const id = s57.id(payload, ID57Length.DEFAULT);
  const hash = s57.hash(payload, H57Length.LEN_256);
  const random = s57.random();

  assert.equal(id.length, 22);
  assert.equal(hash.length, 44);
  assert.equal(random.length, 22);

  const aad = enc.encode('s57-e2e');
  const sealed = s57.encrypt(payload, aad);
  const opened = s57.decrypt(sealed, aad);

  assert.equal(dec.decode(opened), 'e2e-s57-payload');
});

File: implementations/dart/test/s57_e2e_test.dart
import 'package:test/test.dart';
import 'package:f57/b57.dart';

void main() {
  test('S57 e2e secure composition flow', () async {
    final s57 = S57(
      S57Config(
        serverSecretKey: 'S57_SERVER_SECRET_KEY_MUST_BE_LONG_1234567890'.codeUnits,
        environmentSalt: 'staging-v1'.codeUnits,
        keyId: 12,
      ),
    );

    final payload = 'e2e-s57-payload'.codeUnits;

    final id = s57.id(payload, ID57Length.def);
    final hash = s57.hash(payload, H57Length.len256);
    final random = s57.random();

    expect(id.length, 22);
    expect(hash.length, 44);
    expect(random.length, 22);

    final aad = 's57-e2e'.codeUnits;
    final sealed = await s57.encrypt(payload, aad);
    final opened = await s57.decrypt(sealed, aad);

    expect(String.fromCharCodes(opened), 'e2e-s57-payload');
  });
}
